app/main.py

```diff
@@ -1,6 +1,39 @@
-from typing import Callable
+from functools import wraps
+from typing import Callable, Dict, Any
+
+
+def is_immutable(value: Any) -> bool:
+    """Check if a value is immutable."""
+    return isinstance(value, (int, float, str, tuple, frozenset, bool, type(None)))
 
 
 def cache(func: Callable) -> Callable:
-    # Write your code here
-    pass
+    cache_storage: Dict[tuple, Any] = {}
+
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        for arg in args:
+            if not is_immutable(arg):
+                raise TypeError(f"Mutable argument {arg} of type {type(arg)} is not allowed for caching.")
+
+        for key, value in kwargs.items():
+            if not is_immutable(value):
+                raise TypeError(f"Mutable argument {key}={value} of type {type(value)} is not allowed for caching.")
+
+        cache_key = (args, tuple(sorted(kwargs.items())))
```

Comment:
The current method of generating cache keys by sorting keyword arguments may cause issues if the keys are not comparable. Consider using a more robust method to handle keyword arguments, such as using a frozenset for the items.